cos

```diff
@@ -17,58 +39,170 @@ build()
 
 clean()
 {
-    echo "[cos executing] make -C clean"
-    make -C src clean
-}
-
-initialize()
-{
-    echo "[cos executing] make -C src config init"
-    make -C src config init
-}
-
-usage()
-{
-    echo "Usage: " $0 " init|build|reset|compose <script> <output name>|run <binary>"
-    exit 2
+    echo "[cos executing] make -C src clean"
+    make -C src distclean
 }
 
 compose()
 {
     if [ -z "$script" ] || [ -z "$name" ]; then
-	usage
+	    usage
     fi
+
     if ! [ -e "src/composer/target/debug/compose" ]; then
-	echo "Must \"cos build\" before composition. Could not find src/composer/target/debug/compose"
-	exit 2
+        echo "Must \"cos build\" before composition. Could not find src/composer/target/debug/compose"
+        exit 1
     fi
+
     echo "[cos executing] src/composer/target/debug/compose $script $name"
     src/composer/target/debug/compose $script $name
+
+    echo ""
+    echo "[cos generating ISO image]"
+    local dir="system_binaries/cos_build-${name}"
+
+    #detect grub-mkrescue/grub2-mkrescue
+    set +e
+    grub=$(command -v grub-mkrescue)
+    grub2=$(command -v grub2-mkrescue)
+    set -e
+    local grub_command=""
+    if [ "${grub}" != "" ]
+    then
+        grub_command="grub-mkrescue"
+    elif [ "${grub2}" != "" ]
+    then
+        grub_command="grub2-mkrescue"
+    else
+        echo "Cannot find grub-mkrescue/grub2-mkrescue to generate ISO image."
+        exit 1
+    fi
+
+    cd ${dir}
+    echo "set timeout=0" > grub.cfg
+    echo "set default=0" >> grub.cfg
+    echo "menuentry "composite" {" >> grub.cfg
+    echo "  multiboot2 /boot/cos.img" >> grub.cfg
+    echo "}" >> grub.cfg
+
+    mkdir -p iso/boot/grub
+    cp grub.cfg iso/boot/grub/
+    cp cos.img iso/boot/
+    ${grub_command} -d /usr/lib/grub/i386-pc -o ${name}.iso iso
+
+    #clean files
+    rm -rf iso grub.cfg
+
+    echo "Successfully generate ISO image: ${dir}/${name}.iso"
 }
 
 run()
```

Comment:
I think it might be better if we make these a run.sh script in tools dir, just as what we used to have in loader branch

Reply:
From a user's point of view, I think he/she would probably more want to just use the `cos` to finish init->build->compile->run, moving the run logic into another file will confuse the user. Also, the run logic is written into a simple function, thus I think this is OK and easy to understand/maintain.

Reply:
It is more complicated than this. If we ever want people to be able to build the iso with other toolchains, or to do so manually, having a script enables them to do that. Having a script means that we can replace the script more easily per architecture. Of course, the cos script would just call that script. So I don't think that the design comes only down to if it is a simple function or not.

Reply:
Grammar error "Successfully generate..." -> "Successfully generated..."

Reply:
Fixed